common/lib/plugins/bluegreen/blue_green_status_provider.ts

```diff
@@ -0,0 +1,939 @@
+      BlueGreenRole.TARGET,
+      this.bgdId,
+      this.pluginService.getCurrentHostInfo(),
+      this.pluginService,
+      this.getMonitoringProperties(),
+      this.statusCheckIntervalMap,
+      { onBlueGreenStatusChanged: (role, status) => this.prepareStatus(role, status) }
+    );
+  }
+
+  protected getMonitoringProperties(): Map<string, any> {
+    const monitoringConnProperties: Map<string, any> = new Map(this.properties);
+
+    for (const key of monitoringConnProperties.keys()) {
+      if (!key.startsWith(BlueGreenStatusProvider.MONITORING_PROPERTY_PREFIX)) {
+        continue;
+      }
+
+      monitoringConnProperties.delete(key);
+    }
+
+    if (!monitoringConnProperties.has(WrapperProperties.WRAPPER_CONNECT_TIMEOUT.name)) {
+      WrapperProperties.WRAPPER_CONNECT_TIMEOUT.set(monitoringConnProperties, BlueGreenStatusProvider.DEFAULT_CONNECT_TIMEOUT_MS);
+    }
+
+    if (!monitoringConnProperties.has(WrapperProperties.WRAPPER_QUERY_TIMEOUT.name)) {
+      WrapperProperties.WRAPPER_QUERY_TIMEOUT.set(monitoringConnProperties, BlueGreenStatusProvider.DEFAULT_QUERY_TIMEOUT_MS);
+    }
+
+    return monitoringConnProperties;
+  }
+
+  protected isBlueGreenDialect(dialect: any): dialect is BlueGreenDialect {
+    return dialect;
+  }
+
+  protected async prepareStatus(role: BlueGreenRole, interimStatus: BlueGreenInterimStatus): Promise<void> {
+    // Detect changes
+    const contextHash: number = this.getContextHash();
+
+    const storedStatus = this.interimStatuses[role.value];
+
+    if (_.isEqual(storedStatus, interimStatus) && this.lastContextHash === contextHash) {
+      // no changes detected
+      logger.debug(`no changes detected for role: ${role.name}`);
+      return;
+    }
+
+    // There are some changes detected. Let's update summary status.
+    logger.debug(Messages.get("Bgd.interimStatus", this.bgdId, role.name, interimStatus.toString()));
+
+    this.updatePhase(role, interimStatus);
+
+    // Store interimStatus and corresponding hash
+    this.interimStatuses[role.value] = interimStatus;
+    this.lastContextHash = contextHash;
+
+    // Update map of IP addresses.
+    for (const [host, ip] of interimStatus.startIpAddressesByHostMap) {
+      this.hostIpAddresses.set(host, ip);
+    }
+
+    // Update roleByHost based on provided host names.
+    interimStatus.hostNames.forEach((x) => this.roleByHost.set(x.toLowerCase(), role));
+
+    this.updateCorrespondingHosts();
+    this.updateSummaryStatus(role, interimStatus);
+    await this.updateMonitors();
+    this.updateStatusCache();
+    this.logCurrentContext();
+
+    // Log final switchover results.
+    this.logSwitchoverFinalSummary();
+
+    this.resetContextWhenCompleted();
+  }
+
+  protected updatePhase(role: BlueGreenRole, interimStatus: BlueGreenInterimStatus): void {
+    const latestInterimPhase: BlueGreenPhase = !this.interimStatuses[role.value]
+      ? BlueGreenPhase.NOT_CREATED
+      : this.interimStatuses[role.value].blueGreenPhase;
```

Comment:
Not sure if this would ever actually be a problem in practice but should we store this.interimStatuses[role.value] as a local variable so that we are only accessing it once? I'm wondering if there is a chance that this.interimStatuses[role.value] is not null when we check it the first time, and then becomes null before we call `this.interimStatuses[role.value].blueGreenPhase`, in which case you would be dereferencing a null value
```suggestion
    const latestInterimPhase: BlueGreenPhase = !this.interimStatuses[role.value]
      ? BlueGreenPhase.NOT_CREATED
      : this.interimStatuses[role.value].blueGreenPhase;
```